File: sudoku-ums/ums-boot/src/main/java/com/zumin/sudoku/ums/controller/ProfileController.java
package com.zumin.sudoku.ums.controller;

import com.zumin.sudoku.common.alicloud.pojo.OSSCallbackParameter;
import com.zumin.sudoku.common.alicloud.pojo.OSSCallbackResult;
import com.zumin.sudoku.common.alicloud.pojo.OSSPolicy;
import com.zumin.sudoku.common.alicloud.service.OSSService;
import com.zumin.sudoku.common.core.enums.CommonStatusCode;
import com.zumin.sudoku.common.core.exception.BaseException;
import com.zumin.sudoku.common.web.annotation.ComRestController;
import com.zumin.sudoku.common.web.utils.SecurityUtils;
import com.zumin.sudoku.ums.config.ProfileConfig;
import com.zumin.sudoku.ums.config.ProfileConfig.AvatarProperties;
import com.zumin.sudoku.ums.exception.UserException;
import com.zumin.sudoku.ums.service.SysUserService;
import io.swagger.annotations.ApiOperation;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.PostMapping;

@RequiredArgsConstructor
@ComRestController(path = "/profile", tags = "用户档案API接口")
public class ProfileController {

  private final SysUserService userService;

  private final OSSService ossService;

  private final ProfileConfig profileConfig;

  @PostMapping("/uploadAvatarPolicy")
  @ApiOperation("获取上传头像的凭证")
  public OSSPolicy getUploadAvatarPolicy() {
    AvatarProperties avatar = profileConfig.getAvatar();
    return ossService.policy(avatar.getMaxSize(), avatar.getDir(), new OSSCallbackParameter<>("profile/updateAvatarCallback", SecurityUtils.getUserId()));
  }

  @PostMapping("/updateAvatarCallback")
  @ApiOperation("更新头像的回调接口")
  public void updateAvatar() {
    OSSCallbackResult<Long> oosCallback = ossService.resolveCallbackData(Long.class);
    Long userId = oosCallback.getCallbackData();
    if (userId == null) {
      throw new BaseException(CommonStatusCode.UNAUTHORIZED_ACCESS);
    }
    userService.updateAvatar(oosCallback.getFilePath(), userId);
  }
}

File: sudoku-ums/ums-boot/src/main/java/com/zumin/sudoku/ums/controller/ProfileController.kt
package com.zumin.sudoku.ums.controller

import com.zumin.sudoku.common.alicloud.OSSCallbackParameter
import com.zumin.sudoku.common.alicloud.OSSPolicy
import com.zumin.sudoku.common.alicloud.OSSService
import com.zumin.sudoku.common.core.code.CommonStatusCode
import com.zumin.sudoku.common.core.exception.BaseException
import com.zumin.sudoku.common.web.ComRestController
import com.zumin.sudoku.common.web.utils.getCurrentUserId
import com.zumin.sudoku.ums.config.ProfileProperties
import com.zumin.sudoku.ums.service.SysUserService
import io.swagger.annotations.ApiOperation
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.web.bind.annotation.PostMapping

@EnableConfigurationProperties(ProfileProperties::class)
@ComRestController(path = ["/profile"], tags = ["用户档案API接口"])
class ProfileController(
  private val userService: SysUserService,
  private val ossService: OSSService,
  private val profileProperties: ProfileProperties,
) {

  @ApiOperation("获取上传头像的凭证")
  @PostMapping("/uploadAvatarPolicy")
  fun getUploadAvatarPolicy(): OSSPolicy {
    val (dir, maxSize) = profileProperties.avatar
    return ossService.policy(maxSize, dir, OSSCallbackParameter("profile/updateAvatarCallback", getCurrentUserId()))
  }

  @PostMapping("/updateAvatarCallback")
  @ApiOperation("更新头像的回调接口")
  fun updateAvatar() {
    val (callbackData, filePath) = ossService.resolveCallbackData(Long::class.java)
    val userId = callbackData ?: throw BaseException(CommonStatusCode.UNAUTHORIZED_ACCESS)
    userService.updateAvatar(filePath!!, userId)
  }
}